### PR TITLE
Example/tutorial for WMS and TMS

### PR DESCRIPTION
**Pull request by IvanSanchez**
Part of #4239

### CODE REVIEW

File: examples/wms/wms-example1.html [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	<meta charset="utf-8" />
+
+	<meta name="viewport" content="width=device-width, initial-scale=1.0">
+
+	<link rel="stylesheet" href="https://leafletjs-cdn.s3.amazonaws.com/content/leaflet/master/leaflet.css" />
+</head>
+<body>
+	<div id="map" style="width: 600px; height: 400px"></div>
+
+	<script src="https://leafletjs-cdn.s3.amazonaws.com/content/leaflet/master/leaflet.js"></script>
```
Comment (jieter) [inline examples/wms/wms-example1.html]: Sure you want to link to master here?

File: examples/wms/wms-example1.html [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+	<script>
+
+		var map = L.map('map', {
+			center: [-17, -67],
+			zoom: 3
+		});
+
+		var wmsLayer = L.tileLayer.wms('http://demo.opengeo.org/geoserver/ows?', {
```
Comment (jieter) [inline examples/wms/wms-example1.html]: Doesn't really need to be assigned, but maybe doesn't hurt either.

File: examples/wms/wms-example1.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+			layers: 'ne:ne'
+		}).addTo(map);
+
+
+
```
Comment (jieter) [inline examples/wms/wms-example1.html]: Maybe clean up some whitespace

File: examples/wms/wms-example1.html (hunk 1 above)
Comment (IvanSanchez) [inline examples/wms/wms-example1.html]: Let me set change everything to some 1.0.0-beta URLs

File: examples/wms/wms.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+WMS, short for [*web map service*](https://en.wikipedia.org/wiki/Web_Map_Service), is a popular way of publishing maps by professional GIS software (and seldomly used by non-GISers). This format is similar to map tiles, but more generic and not so well optimized for use in web maps. TMS stands for *tiled map service*, and is a WMS-like standard more focused on web maps.
+
+## WMS in Leaflet
+
+When somebody publishes a WMS service, most likely they link to something called a `getCapabilities` document. For this tutorial, we'll use the demo map services from GeoServer, at http://demo.opengeo.org/geoserver/web/. As you can see in that page, "WMS" links to the following URL:
+
+	http://demo.opengeo.org/geoserver/ows?service=wms&version=1.3.0&request=GetCapabilities
+
+Leaflet does not understand WMS GetCapabilities documents. Instead, you have to create a `L.TileLayer.WMS` layer, provide the base WMS URL, and specify whatever WMS options you need.
```
Comment (jieter) [inline examples/wms/wms.md]: Different capitalization here for getCapabilities. Also no backticks around it.

File: examples/wms/wms.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+And the way to use that in a Leaflet map is simply:
+
+	var map = L.map(mapDiv, mapOptions);
+
+	var wmsLayer = L.tileLayer.wms('http://demo.opengeo.org/geoserver/ows?', wmsOptions).addTo(map);
+
+An instance of `L.TileLayer.WMS` needs at least one option: `layers`. Be careful, as the concept of "layer" in Leaflet is different from the concept of "layer" in WMS!
+
+WMS servers define a set of *layers* in the service. While these are defined in the `getCapabilities` document, it's simpler to use software such as [QGIS to see what layers are available in a WMS server](http://www.qgistutorials.com/en/docs/working_with_wms.html):
```
Comment (jieter) [inline examples/wms/wms.md]: might mention that a readable XML is returned but that QGIS makes it more readable.

File: examples/wms/wms.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		'Countries, then boundaries': L.tileLayer.wms('http://demo.opengeo.org/geoserver/ows?', {
+			layers: 'ne:ne_10m_admin_0_countries,ne:ne_10m_admin_0_boundary_lines_land'
+		}),
+
+		'Boundaries, then countries': L.tileLayer.wms('http://demo.opengeo.org/geoserver/ows?', {
+			layers: 'ne:ne_10m_admin_0_boundary_lines_land,ne:ne_10m_admin_0_countries'
+		})
+	};
+
+	L.control.layers(basemaps).addTo(map);
```
Comment (jieter) [inline examples/wms/wms.md]: The stand alone example has
```
L.control.layers(basemaps, {}, {collapsed: false}).addTo(map);
```
here

Comment (IvanSanchez) [inline examples/wms/wms.md]: Yeah, I'll leave it like that, to make the code in the tutorial simpler. How the layers control works is detailed in its own tutorial.

**Comment by IvanSanchez** [comment]
After some proof-reading from @jieter, I've created better templates for the iframed examples - this allows to develop small examples and change the Leaflet version in all framed examples at once.

**Comment by jieter** [comment]
Templates look good!

File: examples/wms/wms.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	var wmsLayer = L.tileLayer.wms('http://demo.opengeo.org/geoserver/ows?', {
+		layers: 'nasa:bluemarble'
+	}).addTo(map);
+
+{% include frame.html url="wms-example2.html" %}
+
+
+The `layers` option is a comma-separated list of layers. If a WMS service has defined several layers, then a request for a map image can refer to more than one layer.
+
+For the example WMS server we're using, there is a `ne:ne_10m_admin_0_countries` WMS layer showing country landmasses and country names, and a `ne:ne_10m_admin_0_boundary_lines_land` WMS layer showing country boundaries. The WMS server will compose both layers in one image if we request both, separated with a comma:
```
Comment (jieter) [inline examples/wms/wms.md]: seems to be an 'in' missing in this sentence.

File: examples/wms/wms.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+	var countriesAndBoundaries = L.tileLayer.wms('http://demo.opengeo.org/geoserver/ows?', {
+		layers: 'ne:ne_10m_admin_0_countries,ne:ne_10m_admin_0_boundary_lines_land'
+	}).addTo(map);
+
+Note this will request *one* image to the WMS server. This is different than creating a `L.TileLayer.WMS` for the countries, another one for the boundaries, and adding them both to the map. In the first case, there is one image request and it's the WMS server who decides how to compose (put on top of each other) the image. In the second case, there would be two image requests and it's the Leaflet webpage who decides how to compose them.
```
Comment (jieter) [inline examples/wms/wms.md]: 'Leaflet webpage' sounds like http://leafletjs.com

File: examples/wms/wms.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+{% include frame.html url="wms-example3.html" %}
+
+
+### Notes to GIS users of WMS services
+
+From a GIS point of view, WMS handling in Leaflet is quite basic. There's no `GetCapabilities` support, no legend support, and no `GetFeatureInfo` support.
+
+`L.TileLayer.WMS` has extra options, which can be found in [Leaflet's API documentation](http://leafletjs.com/reference.html#tilelayer-wms-options). Any option not described there will be passed to the WMS server in the `getImage` URLs.
+
+Also note that Leaflet only supports `CRS:3857`. If your WMS service doesn't serve images in that coordinate system, you might need to use [Proj4Leaflet](https://github.com/kartena/Proj4Leaflet) to use a different coordinate system in Leaflet.
```
Comment (jieter) [inline examples/wms/wms.md]: might want to link to a page explaining EPSG3857, for example http://spatialreference.org/ref/sr-org/6864/

File: examples/wms/wms.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	}).addTo(map);
+
+	var tms_bluemarble = L.tileLayer('http://demo.opengeo.org/geoserver/gwc/service/tms/1.0.0/nasa:bluemarble@EPSG:900913@jpg/{z}/{x}/{y}.jpg', {
+		tms: true
+	});
+
+{% include frame.html url="wms-example4.html" %}
+
+
+A new feature in **Leaflet 1.0** is the ability to use `{-y}` in the URL instead of a `tms: true` option, e.g.:
```
Comment (jieter) [inline examples/wms/wms.md]: yay :+1:

File: examples/wms/wms.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+Using the same OpenGeo WMS/TMS server demo, we can see there's a TMS endpoint at:
+
+	http://demo.opengeo.org/geoserver/gwc/service/tms/1.0.0
+
+Checking the [MapCache help about TMS](http://mapserver.org/mapcache/services.html) and the [TMS specification](https://wiki.osgeo.org/wiki/Tile_Map_Service_Specification) you can see that the URL for a map tile in TMS looks like:
+
+	http://base_url/tms/1.0.0/ {tileset} / {z} / {x} / {y} .png
+
+To use the OpenGeo TMS services as a `L.TileLayer`, we can gheck the capabilities document (the same as the base endpoint, in our case [`http://demo.opengeo.org/geoserver/gwc/service/tms/1.0.0`](http://demo.opengeo.org/geoserver/gwc/service/tms/1.0.0)) to see what `tileset`s are available, and build our base URLs:
```
Comment (jieter) [inline examples/wms/wms.md]: gheck -> check

File: examples/wms/wms.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+{% include frame.html url="wms-example4.html" %}
+
+
+A new feature in **Leaflet 1.0** is the ability to use `{-y}` in the URL instead of a `tms: true` option, e.g.:
+
+	var layer = L.tileLayer('http://base_url/tms/1.0.0/tileset/{z}/{x}/{-y}.png');
+
+The `tms: true` option (in Leaflet 0.7) or `{-y}` (in Leaflet 1.0) are needed because the origin of coordinates of vanilla `L.TileLayer`s is the top left corner, so the Y coordinate goes *down*. In TMS, the origin of coordinates is the *bottom* left corner so the Y coordinate goes *up*.
+
+As you can see, TMS services fit nicely into a `L.TileLayer` definition, so there is no explicit need for them.
```
Comment (perliedman) [inline examples/wms/wms.md]: I'm a bit unsure what "them" refers to in this sentence?

Comment (IvanSanchez) [inline examples/wms/wms.md]: I'm rewording that sentence.

**Pull request merged**